# -*- coding: utf-8 -*-
from setuptools import setup

packages = \
['damona',
 'damona.biocontainers',
 'damona.library',
 'damona.papers',
 'damona.shell',
 'damona.shell.bash',
 'damona.shell.zsh',
 'damona.software',
 'damona.software.bioconvert',
 'damona.software.canu',
 'damona.software.idr',
 'damona.software.multiqc.sequana_0.16.0',
 'damona.software.multiqc.sequana_0.16.0.sequana',
 'damona.software.multiqc.sequana_0.16.0.sequana.multiqc']

package_data = \
{'': ['*'],
 'damona.library': ['R/*', 'conda/*', 'micromamba/*'],
 'damona.shell': ['fish/*'],
 'damona.software': ['art/*',
                     'bamtools/*',
                     'bbtools/*',
                     'bcl2fastq/*',
                     'bedtools/*',
                     'blast/*',
                     'bowtie/*',
                     'bowtie2/*',
                     'busco/*',
                     'bwa/*',
                     'ccs/*',
                     'cd-hit/*',
                     'cellranger_atac/*',
                     'checkm/*',
                     'circlator/*',
                     'falco/*',
                     'fastp/*',
                     'fastqc/*',
                     'flye/*',
                     'freebayes/*',
                     'gffread/*',
                     'graphviz/*',
                     'guppy/*',
                     'gzip/*',
                     'helloworld/*',
                     'hifiasm/*',
                     'hisat2/*',
                     'hmmer/*',
                     'homer/*',
                     'igvtools/*',
                     'ivar/*',
                     'jellyfish/*',
                     'kraken/*',
                     'laa/*',
                     'mafft/*',
                     'medaka/*',
                     'minimap2/*',
                     'multiqc/*',
                     'nanopolish/*',
                     'nextclade/*',
                     'pangolin/*',
                     'pbbam/*',
                     'pbsim/*',
                     'phantompeakqualtools/*',
                     'pigz/*',
                     'polypolish/*',
                     'pplacer/*',
                     'prokka/*',
                     'pycoqc/*',
                     'quast/*',
                     'raxml/*',
                     'rnadiff/*',
                     'rnaseqc/*',
                     'rtools/*',
                     'salmon/*',
                     'samtools/*',
                     'samtools_minimap2/*',
                     'seacr/*',
                     'seqkit/*',
                     'seqtk/*',
                     'sequana/*',
                     'sequana_denovo/*',
                     'sequana_minimal/*',
                     'sequana_perl_tools/*',
                     'sequana_ribofinder/*',
                     'sequana_tools/*',
                     'shustring/*',
                     'snpeff/*',
                     'subread/*',
                     'transdecoder/*',
                     'trf/*',
                     'trim_galore/*',
                     'trinity/*',
                     'trinotate/*',
                     'ucsc/*',
                     'unicycler/*',
                     'vt/*']}

install_requires = \
['PyYAML>=6.0.1,<7.0.0',
 'beautifulsoup4>=4.12.2,<5.0.0',
 'click-completion>=0.5.2,<0.6.0',
 'click>=8.1.7,<9.0.0',
 'colorlog>=6.8.0,<7.0.0',
 'easydev>=0.12.1,<0.13.0',
 'packaging>=23.2,<24.0',
 'requests>=2.31.0,<3.0.0',
 'rich-click>=1.7.2,<2.0.0',
 'spython>=0.3.13,<0.4.0',
 'tqdm>=4.66.1,<5.0.0']

entry_points = \
{'console_scripts': ['damona = damona.script:main']}

setup_kwargs = {
    'name': 'damona',
    'version': '0.11.1',
    'description': 'A manager of singularity containers and their singularity recipes (NGS applications)',
    'long_description': 'DAMONA\n######\n\n\n.. image:: https://badge.fury.io/py/damona.svg\n    :target: https://pypi.python.org/pypi/damona\n\n\n.. image:: https://github.com/cokelaer/damona/actions/workflows/main.yml/badge.svg\n   :target: https://github.com/cokelaer/damona/actions/workflows/main.yml\n\n.. image:: https://coveralls.io/repos/github/cokelaer/damona/badge.svg?branch=master\n    :target: https://coveralls.io/github/cokelaer/damona?branch=master\n\n.. image:: http://readthedocs.org/projects/damona/badge/?version=latest\n    :target: http://damona.readthedocs.org/en/latest/?badge=latest\n    :alt: Documentation Status\n\n.. image:: https://zenodo.org/badge/282275608.svg\n   :target: https://zenodo.org/badge/latestdoi/282275608\n\n\n:Python version: Python 3.8, 3.9, 3.10\n:Source: See  `http://github.com/cokelaer/damona <https://github.com/cokelaer/damona/>`__.\n:Issues: Please fill a report on `github <https://github.com/cokelaer/damona/issues>`__\n:Platform: This is currently only available for Linux distribution with zsh/fish/bash shells (contributions are welcome to port the tool on other platforms/shells)\n\nQuick Start\n===========\n\nAssuming Singularity (Apptainer) is installed on your system, install Damona using **pip** for Python::\n\n    pip install damona\n\nYou need to configure Damona before using it. In a bash shell, type::\n\n    damona\n\nAdd these lines in your .bashrc::\n\n    if [ ! -f  "~/.config/damona/damona.sh" ] ; then\n        source ~/.config/damona/damona.sh\n    fi\n\n**open a new shell** and then use damona::\n\n    damona create TEST\n    damona activate TEST\n    damona install fastqc\n    fastqc\n\nThis should install **fastqc** in the newly created Damona environment (TEST). Type::\n\n    fastqc\n\nit should open a window with the **fastqc** interface.\n\n\nOverview\n========\n\nDamona is a singularity environment manager.\n\nDamona started as a small collections of singularity recipes to help installing third-party tools for\n`Sequana NGS pipelines <https://sequana.readthedocs.io>`_.\n\nDamona is now used in production to create reproducible environments where singularity images and their associated binaries are installed altogether.\n\nIn a nutshell, Damona combines the logic of Conda environments with the\nreproducibility of singularity containers. We believe that it could be useful for\nother projects and therefore decided to release it as an independent tool.\n\n* As of 11st Jna 2024, **Damona** contains 82 software, 128 releases 456 binaries.\n\n\nInstallation\n============\n\nIf you are in a hurry, just type::\n\n    pip install damona --upgrade\n\nYou must install `Singularity <https://sylabs.io/docs>`_ to make use of **Damona**.\n\nIf you are familiar with conda, I believe you can do::\n\n    conda install singularity\n\nType **damona** in a shell. This will initiate the tool with a config file in your HOME/.config/damona directory for bash shell and `fish shell <https://fishshell.com/>`_ users.\n\nBash users should add this code in their ~/.bashrc file::\n\n    source ~/.config/damona/damona.sh\n\nFish shell users should add the following code in their ~/.config/fish/config.fish file::\n\n    source ~/.config/damona/damona.fish\n\nZsh users should add the following code in their ~/.config/fish/config.fish file::\n\n    source ~/.config/damona/damona.zsh\n\nOpen a new shell and you are ready to go. Please see the `Installation in details`_ section for more information.\n\nQuick Start\n===========\n\n**Damona** needs environments to work with.\nFirst, let us *create* one, which is called TEST::\n\n    damona create TEST\n\nSecond, we need to *activate* it. Subsequent insallation will happen in this environment::\n\n    damona activate TEST\n\nFrom there, we can install some binaries/images::\n\n    damona install fastqc:0.11.9\n\nThat\'s it. Time to test. Type **fastqc**.\n\nTo rename this TEST example::\n\n    damona rename TEST --new-name prod\n\nor delete it::\n\n    damona delete prod\n\nSee more examples hereafter or in the user guide on https://damona.readthedocs.io\n\nMotivation\n==========\n\nAs stated on their website, `Conda <https:/docs.conda.io/en/latest>`_ is\nan open source **package** management system\nand **environment** management system.\nConda provides pre-compiled releases of software; they can be installed in\ndifferent local environments that do not interfer with your system. This has\ngreat advantages for developers. For example, you can install a pre-compiled\nlibraries in a minute instead of trying to compile it yourself including all\ndependencies. Different communities have emerged using this\nframework. One of them is `Bioconda <https://bioconda.github.io>`_, which is dedicated to bioinformatics.\n\nAnother great tool that emerged in the last years is\n`Singularity <https://sylabs.io/docs>`_. Singularity containers can be used\nto package entire scientific workflows,\nsoftware and libraries, and even data. It is a simple file that can be shared\nbetween environments and guarantee exectution and reproducibility.\n\nOriginally, Conda provided pre-compiled version of a software. Nowadays, it also provides\na docker and a singularity image of the tool. On the other side, Singularity can include an\nentire conda environment. As you can see everything is there to build reproducible tools and\nenvironment.\n\nNow, what about a software in development that depends on third-party packages ?\nYou would create a conda environment and starts installing the required packages.\nQuickly, you will install another package that will break your environment due\nto unresolved conlicts; this is not common but it happens. In the worst case\nscenario, the environment is broken. In facilities where users depends on you,\nit can be quite stresful and time-consuming to maintain several such\nenvironments. This is why we have moved little by little to a very light conda\nenvironment where known-to-cause-problem packages have been shipped into\nsingularity containers. This means we have to create aliases to those\nsingularities. The singularities can be simple executable containers or full\nenvironment containers with many executables inside. In both cases, one need to\nmanage those containers for different users, pipelines, versions etc. This\nstarted to be cumbersome to have containers in different places and update\nscript that generate the aliases to those executables.\n\n\nThat\'s where **damona** started: we wanted to combine the conda-like environment\nframework to manage our singularity containers more easily.\n\nAlthough **Damona** was started with the `Sequana projet <https://sequana.readthedocs.io>`_,\n**Damona** may be useful for others developers who wish to have a quick and easy\nsolution for their users when they need to install third-party libraries.\n\nBefore showing real-case examples, let us install the software itself and\nunderstand the details.\n\n\n\nInstallation in details\n=======================\n\nThe is the egg and chicken paradox. To get reproducible container with\nsingularity, at some point you need to install singularity itself. That the first\nof the two software that you will need to install. Instructions\nare on `singularity web site <https://sylabs.io/guides/3.6/user-guide/>`_. This\nis not obvious to be honest. You need the GO language to be installed as well. I\npersonally installed from source and it worked like a charm.\n\nSecond, you need **Damona**. This is a pure Python sotfware with only a few\ndependencies. Install it with the **pip** software provided with your Python\ninstallation (Python 3.X)::\n\n    pip install damona --upgrade\n\nType **damona** to create the Damona tree structure. Images and binaries\nwill be saved in your home directory within the\n~/.config/damona directory. There, special files should be available:\n**damona.sh**, **damona.fish**  and **damona.cfg**. Check that those files are present.\n\nFinally, you need to tell your system where to find damona. For bashrc users,\nplease add this line to you bashrc file::\n\n    source ~/.config/damona/damona.sh\n\nopen a new shell and type **damona** and you should be ready to go.\n\nFor fishshell users, please add this line in **~/.config/fish/config.fish***::\n\n    source ~/.config/damona/damona.fish\n\nTutorial\n============\n\nThe **Damona** standalone is called **damona**. It has a documentation that should suffice for most users.\n\nThe main documentation is obtained using::\n\n    damona --help\n\nwhere you will see the list of **Damona** commands (may be different with time) (may be::\n\n\n    activate    Activate a damona environment.\n    clean       Remove orphan images and binaries from all environments.\n    create      Create a new environment\n    deactivate  Deactivate the current Damona environment.\n    delete      Remove an environment\n    env         List all environemnts with some stats.\n    export      Create a bundle of a given environment.\n    info        Print information about a given environment.\n    install     Download and install an image and its binaries.\n    list        List all packages that can be installed\n    remove      Remove binaries or image from an environment.\n    rename      Rename an existing environment\n    search      Search for a container or binary.\n    stats       Get information about Damona images and binaries\n\n\nTo get help for the *install* command, type::\n\n    damona install --help\n\n\n1. *list* available environments\n--------------------------------\n\nBy default you have an environment called **base**. Unlike the **base** environment found in **conda**, it is not\nessential and may be altered. However, it cannot be removed or created. You can check the list of environments using::\n\n    damona env\n\n2. *create* environments\n------------------------\nAll environments are stored in *~/.config/damona/envs/*. You can create a new one as follows::\n\n    damona create TEST\n\nThere, you have a *bin* directory where binaries are going to be installed.\n\nYou can check that it has been created::\n\n    damona env\n\nNote the last line telling you that::\n\n    Your current env is \'TEST\'.\n\n3. activate and deactivate environments\n----------------------------------------\n\nIn order to install new binaries or software package, you must activate an environment. You may activate several but the last one is the *active* one. Let us activate the *TEST* environment::\n\n    damona activate TEST\n\nCheck that it is active using::\n\n    damona env\n\nand look at the last line. It should look like::\n\n    Your current env is \'TEST\'.\n\nWhat is going on when you activate an environment called TEST ? Simple: we append the directory ~/.config/damona/envs/TEST/bin to your PATH where binaries are searched for. This directory is removed when you use the *deactivate* command.\n\n::\n\n    damona deactivate TEST\n    damona env\n\nshould remove the TEST environment from your PATH. You may activate several and deactivate them. In such case, the\nenvironments behave as a Last In First Out principle::\n\n    damona activate base\n    damona activate TEST\n    damona deactivate\n\nRemoves the last activated environments. While this set of commands is more specific::\n\n    damona activate base\n    damona activate TEST\n    damona deactivate base\n\nand keep the TEST environment only in your PATH.\n\n4. **install** a software\n--------------------------\n\nLet us now consider that the TEST environment is active.\n\nDamona provides software that may have several releases. Each software/release comes with binaries that will be\ninstalled together with the underlying singularity image.::\n\n    damona install fastqc:0.11.9\n\nHere, the singularity image corresponding to the release 0.11.9 of the **fastqc** software is downloaded. Then, binaries registered in this release are installed (here the **fastqc** binary only).\n\nAll images are stored in *~/.config/damona/images* and are shared between environments.\n\n\n5. Get **info** about installed images and binaries\n----------------------------------------------------\n\nYou can get the binaries installed in an environment (and the images used by\nthem) using the **info** command::\n\n    damona info TEST\n\n\n6. Search the registry\n------------------------\n\nYou can search for a binary using::\n\n    damona search PATTERN\n\nExternal registry can be set-up. For instance, a damona registry is accessible\nas follows (for demonstration)::\n\n    damona search fastqc --url damona\n\nWhere *damona* is an alias defined in the .config/damona/damona.cfg that\nis set to https://biomics.pasteur.fr/drylab/damona/registry.txt\n\nYou may retrieve images from a website where a registry exists (see the developer\nguide to create a registry yourself).\n\n\n7. combine two different environments\n--------------------------------------\n\nIn damona, you can have sereral environments in parallel and later activate the\none you wish to use. Let us create a new one::\n\n    damone create test1\n\nand check that you now have one more environment::\n\n    damona env\n\nWe want to create an alias to the previously downloaded image of fastqc tool but\nin the *test1* environment. First we activate the newly create environment::\n\n    damona activate test1\n\nthen, we install the container::\n\n    damona install fastqc:0.11.9\n\nThis will not download the image again. It will just create a binary in the\n~/.config/damona/envs/test1/bin directory.\n\nyou can combine this new environment with the base one::\n\n    damona activate base\n\nIf you are interested to know more, please see the User Guide and Developer\nguide here below.\n\n\n\nChangelog\n=========\n\nFrom version 0.10 onwards, we will not mention the new software and their versions\nbut only changes made to the code itself.\n\n========= ========================================================================\nVersion   Description\n========= ========================================================================\n0.11.1    * add pbsim. update hifiasm\n0.11.0    * add precommit, update to use pyproject\n0.10.1    * Fix the get_stats_software wrt new  zenodo API\n0.10.0    * ADD zsh support\n          * UPDATE flye 2.9.1\n          * ADD nanopolish\n          * UPDATE remove nanopolish from sequana_tools binaries\n0.9.1     * ADD hmmer 3.2.2\n          * ADD trinotate 4.0.1\n          * ADD transdecode 5.7.0\n          * ADD trinity 2.15.1\n          * UPDATE bioconvert 1.1.0\n          * update bowtie2 2.5.1\n0.9.0     * refactorise the command \'env\' by splitting into dedicated subcommands\n            create, delete, rename. add progress bar when downloading container\n          * NEW micromamba image to work as a localimage\n          * NEW sequana_minimal package to hold common tools (bwa, samtools,\n            kraken, etc)\n          * NEW ivarm pangolin, nextclade, subread, mafft packages\n          * UPDATE fastp to 0.23.3, gffread to 0.12.7 (3 times lighter).\n          * UPDATE sequana_tools to use micromamba (30% lighter)\n0.8.4     * fix damona stats command to return unique binaries\n          * more recipes and version (e.g. fastqc 0.12.1, graphviz update, etc)\n0.8.3     * create registry specifically for the sandbox (for testing)\n          * add damona community in the uploads\n          * add pbbam, bioconvert, busco, canu, ccs\n          * add polypolish, samtools 1.16.1, sequana 0.14.6, flye 2.9, canu 2.1.1\n            circlator 1.5.5, hifiasm\n0.8.2     * add idr, samtools, homer, bamtools, bedtools, sequana_denovo\n          * add seqkit recipe and container\n          * add shustring\n0.8.1     * Include ability to interact with biocontainers by allowing retrieval\n            of all biocontainers docker images using this syntax:\n            \'damona install biocontainers/xx:1.2.3\' Note that although 9000\n            containers are available, in practice, only about 1000 dockers are\n            on dockerhub, which is already nice :-)\n          *\n0.8.0     * Fix regression to install a software with its version\n0.7.1     * Implement the fish shell\n          * add command "damona list"\n          * rename recipes/ directory into software/ and created a new library/\n            directory for images used as library, that are not installed.\n0.7.0     * Check that singularity is installed\n          * implement the remove command\n            https://github.com/cokelaer/damona/issues/15\n          * more recipes cleanup (https://github.com/cokelaer/damona/issues/12)\n          * removed damona recipes (pure python package)\n          * cleanup all recipes\n          * add zenodo stats (for admin)\n0.6.0     * add ability to upload images on zenodo. No need for external\n            repositories.\n          * ability to add/delete a software from different images\n          * implement --help for the activate/deactivate (non trivial)\n          * add --rename option in \'damona env\'\n          * \'base\' environment is now at the same level as other environments\n          * better bash script; no need for DAMONA_EXE variable anymore.\n0.5.3     * Fixing config/shell\n0.5.2     * add missing shell package\n0.5.1     * add DAMONA_SINGULARITY_OPTIONS env variable in the binary\n          * Fix the way binaries are found in the releases.\n          * new recipes: rtools\n          * new releases: sequana_tools_0.10.0\n          * Fix shell script to handle DAMONA_EXE variable\n0.5.0     * Major refactoring.\n\n            - Simplification of the registries (dropping notion of exe/set\n              class\n            - Main script should now be fully functional with functional\n              activation/deactivation.\n            - New command to build images from local recipes or dockerhub\n              entries.\n            - Install command can now install local container.\n            - DAMONA_PATH can be set to install damona images/binaries\n              anywhere, not just in local home.\n            - check md5 of images to not download/copy again\n0.4.3     * Implement damona activate/deactivate\n0.4.2     * Fix typo in the creation of aliases for \'set\' containers\n0.4.1     * implemented aliases for the --from-url option stored in a\n            damona.cfg file\n0.4.0     * implemented the \'env\' and \'activate\' command\n          * ability to setup an external registry on any https and retrieve\n            registry from there to download external images\n0.3.X     * add gffread, rnadiff recipes\n0.3.0     * A stable version with documentation and >95% coverage read-yto-use\n0.2.3     * add new recipes (rnadiff)\n0.2.2     * Download latest if no version provided\n          * include *build* command to build image locally\n0.2.1     fixed manifest\n0.2.0     first working version of damona to pull image locally with binaries\n0.1.1     small update to fix RTD, travis, coveralls\n0.1       first release to test feasibility of the project\n========= ========================================================================\n',
    'author': 'Thomas Cokelaer',
    'author_email': 'thomas.cokelaer@pasteur.fr',
    'maintainer': 'None',
    'maintainer_email': 'None',
    'url': 'https://github.com/cokelaer/damona',
    'packages': packages,
    'package_data': package_data,
    'install_requires': install_requires,
    'entry_points': entry_points,
    'python_requires': '>=3.8,<4.0',
}


setup(**setup_kwargs)
